handler/event_handler.go

```diff
@@ -0,0 +1,245 @@
+package handler
+
+import (
+	"bytes"
+	"encoding/json"
+	"errors"
+	"io/ioutil"
+	"net/http"
+	"net/url"
+	"reflect"
+	"time"
+
+	"github.com/ONSdigital/dp-import-reporter/config"
+	"github.com/ONSdigital/go-ns/log"
+	"github.com/coocood/freecache"
+)
+
+const failed = "failed"
+
+//main function that triggers everything else
+func (e *EventReport) HandleEvent(httpClient *http.Client, c *freecache.Cache, cfg *config.Config) error {
+	log.Info("Starting error handle", log.Data{"INSTANCE_ID": e.InstanceID, "ERROR_MSG": e.EventMsg})
+
+	status, events, err := e.checkInstance(httpClient, cfg)
+	if err != nil {
+		return err
+	}
+	log.Info("Successfully checked instance", log.Data{"INSTANCE_ID": e.InstanceID, "INSTANCE_STATE": status})
+	instanceEvents := &InstanceEvent{
+		Type:          e.EventType,
+		Message:       e.EventMsg,
+		MessageOffset: "0",
+	}
+
+	timeNow := time.Now()
+
+	jsonUpload, err := json.Marshal(Event{
+		Type:          e.EventType,
+		Time:          &timeNow,
+		Message:       e.EventMsg,
+		MessageOffset: "0",
+	})
+
+	key := []byte(e.InstanceID)
+	value := []byte(e.EventMsg)
+	expire := 25
+
+	check := arraySlicing(instanceEvents, events)
+	if check {
+		_, err := c.Get(key)
+		if err != nil {
+			c.Set(key, value, expire)
+		} else {
+			c.Del(key)
+		}
+	}
+
+	got, err := c.Get(key)
+	if err != nil {
+		c.Set(key, value, expire)
+		err := e.putEvent(httpClient, jsonUpload, cfg, status)
+		if err != nil {
+			return err
+		}
+		return nil
+	}
+	log.Info("This instance is saved inmemory already.", log.Data{e.InstanceID: "this instance is saved inmemory", "lock": got})
+	return nil
+}
+
+/*this puts an event into the database under the instance you chose
+it does some checks to make sure the instance exists and checks the status
+if the status isn't already failed it will turn that instance to failed */
+func (e *EventReport) putEvent(httpClient *http.Client, json []byte, cfg *config.Config, status string) error {
```

Comment:
Method name putEvent but the apiRequest argument is "POST" - potential inconsistency